pkg/processor/build/builder.go

```diff
@@ -1552,6 +1564,71 @@ func (b *Builder) getS3FunctionItemKey() (string, error) {
 	return s3Attributes["s3ItemKey"], nil
 }
 
+func (b *Builder) resolveGitReference() (string, error) {
+	if b.options.FunctionConfig.Spec.Build.CodeEntryAttributes["gitBranch"] != nil {
+		return fmt.Sprintf("refs/heads/%s", b.options.FunctionConfig.Spec.Build.CodeEntryAttributes["gitBranch"]), nil
+	}
+	if b.options.FunctionConfig.Spec.Build.CodeEntryAttributes["gitTag"] != nil {
+		return fmt.Sprintf("refs/tags/%s", b.options.FunctionConfig.Spec.Build.CodeEntryAttributes["gitTag"]), nil
+	}
+	if b.options.FunctionConfig.Spec.Build.CodeEntryAttributes["gitReference"] != nil {
+		return b.options.FunctionConfig.Spec.Build.CodeEntryAttributes["gitReference"].(string), nil
+	}
+
+	return "", errors.New("No git reference was specified. (must specify gitBranch/gitTag/gitReference)")
+}
+
+func (b *Builder) parseFunctionGitCredentials() *githttp.BasicAuth {
+	var username, password string
+
+	_, found := b.options.FunctionConfig.Spec.Build.CodeEntryAttributes["gitUsername"]
+	if found {
+		username = b.options.FunctionConfig.Spec.Build.CodeEntryAttributes["gitUsername"].(string)
+	}
+
+	_, found = b.options.FunctionConfig.Spec.Build.CodeEntryAttributes["gitPassword"]
+	if found {
+		password = b.options.FunctionConfig.Spec.Build.CodeEntryAttributes["gitPassword"].(string)
+	}
+
+	if username != "" || password != "" {
+		return &githttp.BasicAuth{
+			Username: username,
+			Password: password,
+		}
+	}
+
+	return nil
+}
+
+func (b *Builder) cloneFunctionFromGit(tempDir, functionPath string) error {
```

Comment:
move outside of builder, should be `processor/build/util/git.go` or something similar

Reply:
why? this function consists only of resolving credentials/reference (that won't be shared with anything else - as it resolves it from CodeEntryAttributes) and using `go-git`'s PlainClone function